File: frameworks/python/sanic/Implementation/routes/query.py
from sanic import Blueprint
from sanic.response import json
from sanic_ext import openapi, validate

from answers import echoed
from documented import Payload, answers, query_of, refuses
from models import Camel
from payloads import Payloads


class Page(Camel):
    page: int


@openapi.component
class Search(Camel):
    """query.many's eight values, which forms.urlencoded posts as a form."""

    page: int
    size: int
    status: str
    category: str
    sort: str
    q: str
    min_price: int
    max_price: int


class EchoedPage(Payload):
    echo: Page


class EchoedSearch(Payload):
    echo: Search


def blueprint(p: Payloads) -> Blueprint:
    """query: the query string bound to a model by sanic-ext's @validate, which converts each value
    to the type the model declares, so the model is already the echo."""
    routes = Blueprint("query")

    @routes.get("/query/one")
    @query_of(Page)
    @answers(EchoedPage, "items.small, with the page")
    @refuses(400, "sanic-ext's refusal of a value the model cannot bind")
    @validate(query=Page)
    async def one(request, query: Page):
        return json(echoed(p.small, query.model_dump()))

    @routes.get("/query/many")
    @query_of(Search)
    @answers(EchoedSearch, "items.small, with the eight values")
    @refuses(400, "sanic-ext's refusal of a value the model cannot bind")
    @validate(query=Search)
    async def many(request, query: Search):
        return json(echoed(p.small, query.model_dump()))

    return routes

File: frameworks/go/huma/Implementation/query.go
package implementation

import (
	"context"

	"github.com/danielgtaylor/huma/v2"
)

type QueryOne struct {
	Page int `query:"page" json:"page"`
}

// Search is query.many's eight values, which forms.urlencoded posts as a form.
type Search struct {
	Page     int    `query:"page" json:"page"`
	Size     int    `query:"size" json:"size"`
	Status   string `query:"status" json:"status"`
	Category string `query:"category" json:"category"`
	Sort     string `query:"sort" json:"sort"`
	Q        string `query:"q" json:"q"`
	MinPrice int    `query:"minPrice" json:"minPrice"`
	MaxPrice int    `query:"maxPrice" json:"maxPrice"`
}

// queryRoutes bind the query string into typed input fields, which Huma converts.
func queryRoutes(api huma.API, p *Payloads) {
	huma.Get(api, "/query/one", func(ctx context.Context, in *QueryOne) (*EchoedOutput[QueryOne], error) {
		return &EchoedOutput[QueryOne]{Body: Echoed[QueryOne]{&p.Small, *in}}, nil
	})

	huma.Get(api, "/query/many", func(ctx context.Context, in *Search) (*EchoedOutput[Search], error) {
		return &EchoedOutput[Search]{Body: Echoed[Search]{&p.Small, *in}}, nil
	})
}
